app/uk/gov/hmrc/gform/gformbackend/GformConnector.scala

```diff
@@ -119,12 +119,31 @@ class GformConnector(ws: WSHttp, baseUrl: String) {
             }
 
           }
-          .recover {
+          .recoverWith {
             case UpstreamErrorResponse.WithStatusCode(statusCode, _) if statusCode == StatusCodes.NotFound.intValue =>
               logger.info(
                 s"Attempt to access form $formIdData, but form not found in MongoDB, attempt to look for $formIdDataOriginal as a fallback failed."
               )
-              None
+              def getFirstTemplateByLegacyIds(legacyIds: List[FormTemplateId]): Future[Option[FormTemplate]] =
+                legacyIds match {
+                  case Nil => Future.successful(None)
+                  case legacyFormId :: tail =>
+                    getFormTemplate(legacyFormId).map(Some(_)).recoverWith {
+                      case UpstreamErrorResponse.WithStatusCode(statusCode, _)
+                          if statusCode == StatusCodes.NotFound.intValue =>
+                        getFirstTemplateByLegacyIds(tail)
+                    }
+                }
+
+              val legacyFormIds = formTemplate.legacyFormIds.map(_.toList).getOrElse(Nil)
+
+              OptionT(getFirstTemplateByLegacyIds(legacyFormIds)).flatMap { formTemplate =>
+                val updatedFormIdData = formIdData.withTemplateId(formTemplate)
+                OptionT(maybeForm(updatedFormIdData, formTemplate)).semiflatTap { form =>
+                  createFormFromLegacy(updatedFormIdData, formIdData)
+                  Future.successful(form.copy(_id = formIdData.toFormId, formTemplateId = formTemplate._id))
```

Comment:
This looks brittle. Parallel computation where "hard to reason about things" can happen. Let's make `createFormFromLegacy` to return `Form` which will be returned as a result.